src/main/java/org/broadinstitute/hellbender/metrics/InsertSizeMetricsCollector.java

```diff
@@ -53,135 +69,154 @@ protected InsertSizeMetricsCollectorArgs makeArg(SAMRecord samRecord, ReferenceS
+     */
+    public PerUnitInsertSizeMetricsCollector combineUnit(
+            PerUnitInsertSizeMetricsCollector collector1,
+            PerUnitInsertSizeMetricsCollector collector2) {
+        return collector1.combine(collector2);
+    }
+
+    /**
+     * Finish the metrics collection by saving any results to a metrics file.
+     * @param metricsFile a metricsFile where the collected metrics should be stored
+     * @param inputName the name of the input, for optional inclusion in the metrics file
+     * @param authHolder
+     */
+    public void finish(
+            final MetricsFile<InsertSizeMetrics, Integer> metricsFile,
+            final String inputName,
+            final AuthHolder authHolder)
+    {
+        super.finish();
+        addAllLevelsToFile(metricsFile);
+
+        if (metricsFile.getNumHistograms() == 0) { // can happen if user sets MINIMUM_PCT = 0.5, etc.
+            log.warn("All data categories were discarded because they contained < " + inputArgs.minimumPct +
+                     " of the total aligned paired data.");
+            final PerUnitInsertSizeMetricsCollector allReadsCollector = getAllReadsCollector();
+            log.warn("Total mapped pairs in all categories: " + (allReadsCollector == null ?
+                    allReadsCollector :
+                    allReadsCollector.getTotalInserts()));
+        }
+        else {
+            if (null != authHolder) {
+                MetricsUtils.saveMetrics(metricsFile, inputArgs.output, authHolder);
             }
             else {
-                prefix = "All_Reads.";
+                metricsFile.write(new File(inputArgs.output));
+            }
+            if (inputArgs.producePlot) {
+                writeHistogramPDF(inputName);
             }
-            Histograms.put(SamPairUtil.PairOrientation.FR,     new Histogram<>("insert_size", prefix + "fr_count"));
-            Histograms.put(SamPairUtil.PairOrientation.TANDEM, new Histogram<>("insert_size", prefix + "tandem_count"));
-            Histograms.put(SamPairUtil.PairOrientation.RF,     new Histogram<>("insert_size", prefix + "rf_count"));
         }
+    }
 
-        public void acceptRecord(final InsertSizeMetricsCollectorArgs args) {
-            Histograms.get(args.getPairOrientation()).increment(args.getInsertSize());
-        }
+    /**
+     * Calls R script to plot histogram(s) in PDF.
+     */
+    private void writeHistogramPDF(final String inputName) throws RScriptExecutorException {
+        // path to Picard R script for producing histograms in PDF files.
+        final String R_SCRIPT = "insertSizeHistogram.R";
+
+        File histFile = new File(inputArgs.histogramPlotFile);
+        IOUtil.assertFileIsWritable(histFile);
+
+        final RScriptExecutor executor = new RScriptExecutor();
+        executor.addScript(new Resource(R_SCRIPT,
+                org.broadinstitute.hellbender.tools.picard.analysis.CollectInsertSizeMetrics.class));
```

Comment:
Good point - I moved the R file to a metrics resource folder, and no longer need the qualified name.